src/h2_stream_set.erl

```diff
@@ -0,0 +1,855 @@
+                    %% process, or it will just hang out there.
+                    http2_stream:stop(Pid),
+                    {error, ?REFUSED_STREAM, #closed_stream{id=StreamId}};
+                NewStreamSet ->
+                    NewStreamSet
+            end
+    end.
+
+-spec get_peer_subset(
+        stream_id(),
+        stream_set()) ->
+                               peer_subset().
+get_peer_subset(Id, StreamSet) ->
+    case {Id rem 2, StreamSet#stream_set.type} of
+        {0, client} ->
+            StreamSet#stream_set.theirs;
+        {1, client} ->
+            StreamSet#stream_set.mine;
+        {0, server} ->
+            StreamSet#stream_set.mine;
+        {1, server} ->
+            StreamSet#stream_set.theirs
+    end.
+
+-spec set_peer_subset(
+        Id :: stream_id(),
+        StreamSet :: stream_set(),
+        NewPeerSubset :: peer_subset()
+                         ) ->
+                             stream_set().
+set_peer_subset(Id, StreamSet, NewPeerSubset) ->
+    case {Id rem 2, StreamSet#stream_set.type} of
+        {0, client} ->
+            StreamSet#stream_set{
+              theirs=NewPeerSubset
+             };
+        {1, client} ->
+            StreamSet#stream_set{
+              mine=NewPeerSubset
+             };
+        {0, server} ->
+            StreamSet#stream_set{
+              mine=NewPeerSubset
+             };
+        {1, server} ->
+            StreamSet#stream_set{
+              theirs=NewPeerSubset
+             }
+    end.
+
+%% get/2 gets a stream. The logic in here basically just chooses which
+%% subset.
+-spec get(Id :: stream_id(),
+          Streams :: stream_set()) ->
+                 stream().
+get(Id, StreamSet) ->
+    get_from_subset(Id,
+                    get_peer_subset(
+                      Id,
+                      StreamSet)).
+
+-spec get_from_subset(
+        Id :: stream_id(),
+        PeerSubset :: peer_subset())
+                     ->
+                             stream().
+get_from_subset(Id,
+                #peer_subset{
+                   lowest_stream_id=Lowest
+                  })
+  when Id < Lowest ->
+    #closed_stream{id=Id};
+get_from_subset(Id,
+                #peer_subset{
+                   next_available_stream_id=Next
+                  })
+  when Id >= Next ->
+    #idle_stream{id=Id};
+get_from_subset(Id, PeerSubset) ->
+    case lists:keyfind(Id, 2, PeerSubset#peer_subset.active) of
+        false ->
+            #closed_stream{id=Id};
+        Stream ->
+            Stream
+    end.
+
+-spec upsert(
+        Stream :: stream(),
+        StreamSet :: stream_set()) ->
+                    stream_set()
+                  | {error, error_code()}.
+%% Can't store idle streams
+upsert(#idle_stream{}, StreamSet) ->
+    StreamSet;
+upsert(Stream, StreamSet) ->
+    StreamId = stream_id(Stream),
+    PeerSubset =
+        get_peer_subset(StreamId, StreamSet),
+    case
+        upsert_peer_subset(Stream, PeerSubset)
```

Comment:
these linebreaks feel unnecessary? the lines are well under 80 chars concatenated